### PR TITLE
feat (api, hatchery): book for register

### PR DESCRIPTION
**Pull request by yesnault**
Signed-off-by: Yvonnick Esnault <yvonnick.esnault@corp.ovh.com>

### CODE REVIEW

File: engine/api/worker_model.go [hunk 1]
```diff
@@ -76,6 +76,19 @@ func (api *API) addWorkerModelHandler() Handler {
...
 }
 
+func (api *API) bookWorkerModelHandler() Handler {
+	return func(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+		workerModelID, errr := requestVarInt(r, "permModelID")
+		if errr != nil {
+			return sdk.WrapError(errr, "updateWorkerModel> Invalid permModelID")
+		}
+		if _, err := worker.BookForRegister(api.Cache, workerModelID, getHatchery(ctx)); err != nil {
+			return sdk.WrapError(err, "updateWorkerModel> worker model already booked")
```
Comment (fsamin) [inline engine/api/worker_model.go]: redundant wrap message

File: engine/api/worker_model.go [hunk 2]
```diff
@@ -76,6 +76,19 @@ func (api *API) addWorkerModelHandler() Handler {
...
+		}
+		return WriteJSON(w, r, nil, http.StatusOK)
```
Comment (fsamin) [inline engine/api/worker_model.go]: dont write json if there is nothing to write. 
return nil and i guess the status code will be 204

File: engine/api/worker/model.go [hunk 3]
```diff
@@ -248,3 +250,19 @@ func updateRegistration(db gorp.SqlExecutor, modelID int64) error {
...
+	return cache.Key("book", "workermodel", strconv.FormatInt(id, 10))
+}
+
+// BookForRegister books a worker model for register, used by hatcheries
+func BookForRegister(store cache.Store, id int64, hatchery *sdk.Hatchery) (*sdk.Hatchery, error) {
+	k := keyBookWorkerModel(id)
+	h := sdk.Hatchery{}
+	if !store.Get(k, &h) {
+		// worker model not already booked, book it for 6 min
+		store.SetWithTTL(k, hatchery, 360)
```
Comment (fsamin) [inline engine/api/worker/model.go]: 360 => constant

File: sdk/cdsclient/interface.go [hunk 4]
```diff
@@ -152,24 +152,25 @@ type QueueClient interface {
 
 // TemplateClient exposes queue related functions
 type TemplateClient interface {
+	TemplateApplicationCreate(projectKey, name string, template *sdk.Template) error
```
Comment (fsamin) [inline sdk/cdsclient/interface.go]: ?

File: sdk/hatchery/register.go [hunk 5]
```diff
@@ -230,18 +230,22 @@ func workerRegister(h Interface, models []sdk.Model, nRegister *int64) error {
...
-			log.Info("workerRegister> spawn a worker for register worker model %s (%d)", models[k].Name, models[k].ID)
-			atomic.AddInt64(nRegister, 1)
-			currentRegister++
-			go func(m sdk.Model) {
-				if _, errSpawn := h.SpawnWorker(SpawnArguments{Model: m, IsWorkflowJob: false, JobID: 0, Requirements: nil, RegisterOnly: true, LogInfo: "spawn for register"}); errSpawn != nil {
-					log.Warning("workerRegister> cannot spawn worker for register: %s", m.Name, errSpawn)
-					if err := h.Client().WorkerModelSpawnError(m.ID, fmt.Sprintf("workerRegister> cannot spawn worker for register: %s", errSpawn)); err != nil {
-						log.Error("workerRegister> error on call client.WorkerModelSpawnError on worker model %s for register: %s", m.Name, err)
+			if err := h.Client().WorkerModelBook(models[k].ID); err != nil {
+				log.Error("workerRegister> error on call client.WorkerModelBook on worker model %s for register: %s", models[k].Name, err)
```
Comment (fsamin) [inline sdk/hatchery/register.go]: log message is way too long

File: sdk/hatchery/register.go [hunk 6]
```diff
@@ -230,18 +230,22 @@ func workerRegister(h Interface, models []sdk.Model, nRegister *int64) error {
...
+			} else {
+				log.Info("workerRegister> spawn a worker for register worker model %s (%d)", models[k].Name, models[k].ID)
```
Comment (fsamin) [inline sdk/hatchery/register.go]: this is too long 
"workerRegister> spawning model %s (%d)" should be enough

File: sdk/cdsclient/interface.go (hunk 4 above)
Comment (yesnault) [reply sdk/cdsclient/interface.go]: it's just a alpha re-order

Comment (fsamin) [reply sdk/cdsclient/interface.go]: ok sorry

Comment (fsamin) [reply sdk/cdsclient/interface.go]: ah ok; sorry

File: engine/api/worker_model.go [hunk 7]
```diff
@@ -80,12 +80,11 @@ func (api *API) bookWorkerModelHandler() Handler {
 	return func(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
 		workerModelID, errr := requestVarInt(r, "permModelID")
 		if errr != nil {
-			return sdk.WrapError(errr, "updateWorkerModel> Invalid permModelID")
+			return sdk.WrapError(errr, "bookWorkerModelHandler> Invalid permModelID")
 		}
 		if _, err := worker.BookForRegister(api.Cache, workerModelID, getHatchery(ctx)); err != nil {
-			return sdk.WrapError(err, "updateWorkerModel> worker model already booked")
+			return sdk.WrapError(err, "bookWorkerModelHandler>")
 		}
```
Comment (fsamin) [inline engine/api/worker_model.go]: missing return at end of function

**Pull request merged**